rs/sns/testing/src/utils.rs

```diff
@@ -0,0 +1,221 @@
+        .map(PrincipalId::from)
+}
+
+pub async fn check_canister_installed<C: CallCanisters>(
+    agent: &C,
+    canister_id: &CanisterId,
+) -> bool {
+    agent
+        .canister_info(*canister_id)
+        .await
+        .map(|canister_info| canister_info.module_hash.is_some())
+        .unwrap_or_default()
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum SnsTestingNetworkValidationError {
+    #[error("NNS canister '{0}' is missing")]
+    MissingNnsCanister(String),
+    #[error("Network does not have any SNS subnets")]
+    MissingSnsSubnets,
+    #[error("Failed to get SNS subnet IDs: {0}")]
+    FailedToGetSnsSubnetIds(String),
+    #[error("SNS WASM doesn't have any upgrade steps")]
+    MissingSnsWasmUpgradeSteps,
+    #[error("Failed to get SNS WASM upgrade steps: {0}")]
+    FailedToGetSnsWasmUpgradeSteps(String),
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum SnsTestingCanisterValidationError {
+    #[error("Canister {0} is not installed")]
+    CanisterNotInstalled(CanisterId),
+    #[error("Canister {0} is not controlled by the NNS Root canister")]
+    CanisterNotControlledByNnsRoot(CanisterId),
+    #[error("Failed to get canister '{0}' info: {1}")]
+    FailedToGetCanisterInfo(CanisterId, String),
+}
+
+// Function that validates the provided IC network setup.
+// It ensures that all required NNS canisters exist and are installed
+// and the network has at least one SNS subnet.
+pub async fn validate_network<C: CallCanisters>(
+    agent: &C,
+) -> Vec<SnsTestingNetworkValidationError> {
+    let canisters_installed = join_all(
+        ALL_NNS_CANISTER_IDS
+            .iter()
+            .map(|canister_id| async { check_canister_installed(agent, canister_id).await }),
+    )
+    .await;
+    let (_, validation_errors): (Vec<_>, Vec<_>) = ALL_NNS_CANISTER_IDS
+        .iter()
+        .zip(canisters_installed)
+        .map(|(canister_id, installed)| {
+            if !installed {
+                Err(SnsTestingNetworkValidationError::MissingNnsCanister(
+                    canister_id_to_nns_canister_name(**canister_id),
+                ))
+            } else {
+                Ok(())
+            }
+        })
+        .partition(Result::is_ok);
+    let mut validation_errors: Vec<SnsTestingNetworkValidationError> = validation_errors
+        .into_iter()
+        .map(Result::unwrap_err)
+        .collect();
+    match get_sns_subnet_ids(agent).await {
+        Ok(response) => {
+            if response.sns_subnet_ids.is_empty() {
+                validation_errors.push(SnsTestingNetworkValidationError::MissingSnsSubnets);
+            }
+        }
+        Err(err) => {
+            validation_errors.push(SnsTestingNetworkValidationError::FailedToGetSnsSubnetIds(
+                err.to_string(),
+            ));
+        }
+    }
+
+    match list_upgrade_steps(agent, None, None, 0).await {
+        Ok(response) => {
+            if response.steps.is_empty() {
```

Comment:
This seems like a good approximation. 

I suppose this error is not about upgrade steps, but about the fact that there's not Wasm to deploy an SNS in the first place. Right? 

Maybe then we should check that `response.steps.len() == 6` or something slightly stronger.

Reply:
I think I assumed that a single upgrade could add all modules, but this isn't the case.

`response.steps.len() == 6` sounds alright, but I think that `get_latest_sns_version_pretty().len == 6` would be slightly more robust

Reply:
Both sound fine to me.